### PR TITLE
[PW-6400] - The POST Donations API can be called multiple times

### PR DESCRIPTION
**Pull request by candemiralp**
**Description**

The POST Donations API could be called multiple times from the internal REST endpoint.

'donationToken' in the payment's 'additionalInformation' field is removed after each API call. Therefore, duplicate requests fail since there aren't any corresponding donationToken in the database.

**Tested scenarios**

**Fixed issue**:  PW-6400

### CODE REVIEW

File: Model/Api/AdyenDonations.php [hunk 1]
```diff
@@ -96,6 +107,32 @@ public function donate($payload)
...
+        return $donationResult;
+    }
+
+    private function incrementTryCount($order)
+    {
+        $donationTryCount = $order->getPayment()->getAdditionalInformation('donationTryCount');
+
+        if (!$donationTryCount) {
+            $order->getPayment()->setAdditionalInformation('donationTryCount', 1);
+            $order->save();
```
Comment (acampos1916) [inline Model/Api/AdyenDonations.php:124]: save can be called for both IF and ELSE if moved after the closing bracket

File: Model/Api/AdyenDonations.php [hunk 2]
```diff
@@ -96,6 +107,32 @@ public function donate($payload)
...
+        }
+        else {
+            $donationTryCount += 1;
+            $order->getPayment()->setAdditionalInformation('donationTryCount', $donationTryCount);
```
Comment (acampos1916) [inline Model/Api/AdyenDonations.php:128]: 'donationTryCount' being used multiple times should be a constant

File: Model/Api/AdyenDonations.php [hunk 3]
```diff
@@ -80,10 +80,21 @@ public function donate($payload)
         /** @var Order */
         $order = $this->orderFactory->create()->load($this->checkoutSession->getLastOrderId());
 
+        $this->incrementTryCount($order);
```
Comment (peterojo) [inline Model/Api/AdyenDonations.php:83]: Shouldn't we be saving this `donationTryCount` only if a donation has failed?

Comment (peterojo) [reply Model/Api/AdyenDonations.php:83]: Seems to me if a donation has succeeded we can just delete the token end of story

Comment (candemiralp) [reply Model/Api/AdyenDonations.php:83]: If the donation fails, it just throws an exception in the execute method and doesn't return here. So I decided to count at the beginning to be on the safe side for the failing scenarios.

File: Model/Api/AdyenDonations.php (hunk 2 above)
Comment (candemiralp) [reply Model/Api/AdyenDonations.php]: https://github.com/Adyen/adyen-magento2/pull/1428/commits/b6068376284c73fd88555611dc96e08375aba649

File: Model/Api/AdyenDonations.php (hunk 1 above)
Comment (candemiralp) [reply Model/Api/AdyenDonations.php]: https://github.com/Adyen/adyen-magento2/pull/1428/commits/b6068376284c73fd88555611dc96e08375aba649

File: Model/Api/AdyenDonations.php (hunk 3 above)
Comment (candemiralp) [reply Model/Api/AdyenDonations.php]: https://github.com/Adyen/adyen-magento2/pull/1428/commits/454cf56730d6f05442ed9a0a73d4d8095ee42285

**Review by peterojo — APPROVED** [review]

**Review by acampos1916 — APPROVED** [review]

**Pull request merged**